planner/core/logical_plan_builder.go

```diff
@@ -975,6 +975,21 @@ func (b *PlanBuilder) buildSelection(ctx context.Context, p LogicalPlan, where a
 	if len(cnfExpres) == 0 {
 		return p, nil
 	}
+	// check expr field types.
+	for i, expr := range cnfExpres {
+		if expr.GetType().EvalType() == types.ETString {
+			tp := &types.FieldType{
+				Tp:      mysql.TypeDouble,
+				Flag:    expr.GetType().Flag & mysql.UnsignedFlag,
+				Flen:    mysql.MaxRealWidth,
+				Decimal: types.UnspecifiedLength,
+			}
+			types.SetBinChsClnFlag(tp)
+			if res, ok := expression.TryPushCastDownToControlFunctionForHybridType(b.ctx, expr, tp); ok {
```

Comment:
What if the `expr` is `cast(if as string)` here?

Reply:
It will push cast as string into if function.

Reply:
But the `expr.FuncName` is Cast.

Reply:
I got it wrong...
If the `expr` is `cast(if as string)`,  the `res` will be `cast(if as string)`.